dimod/generators/mimo.py

```diff
@@ -0,0 +1,1058 @@
+    root-power of transmission between integer-indexed transmitters and 
+    receivers.
+
+    It sets uniform transmission of power for on-site transmitter-receiver pairs
+    and uniform transmission from transmitters to receivers up to graph distance 
+    1. 
+    """
+    # Developer note: this code should exploit sparsity, and track the label map.
+    # The code could be generalized to account for asymmetric transmission patterns, 
+    # or real-valued spatial structure.
+
+    if any('num_transmitters' in lattice.nodes[n] for n in lattice.nodes) or \
+       any('num_receivers' in lattice.nodes[n] for n in lattice.nodes):
+
+        node_to_transmitters = {}   #Integer labels of transmitters at node
+        node_to_receivers = {}      #Integer labels of receivers at node
+        t_ind = 0
+        r_ind = 0
+        for n in lattice.nodes:
+            num = transmitters_per_node
+            if 'num_transmitters' in lattice.nodes[n]:
+                num = lattice.nodes[n]['num_transmitters']
+            node_to_transmitters[n] = list(range(t_ind, t_ind + num))
+            t_ind = t_ind + num
+
+            num = receivers_per_node
+            if 'num_receivers' in lattice.nodes[n]:
+                num = lattice.nodes[n]['num_receivers']
+            node_to_receivers[n] = list(range(r_ind, r_ind + num))
+            r_ind = r_ind + num
+
+        A = np.zeros(shape=(r_ind, t_ind))
+        for n0 in lattice.nodes:
+            root_receivers = node_to_receivers[n0]
+            for r in root_receivers:
+                for t in node_to_transmitters[n0]:
+                    A[r, t] = 1
+                for neigh in lattice.neighbors(n0):
+                    for t in node_to_transmitters[neigh]:
+                        A[r, t] = neighbor_root_attenuation
+    else:
+        A = np.identity(lattice.number_of_nodes())
+        # Uniform case:
+        node_to_int = {n:idx for idx, n in enumerate(lattice.nodes())}
+        for n0 in lattice.nodes:
+            root = node_to_int[n0]
+            for neigh in lattice.neighbors(n0):
+                A[node_to_int[neigh], root] = neighbor_root_attenuation
+        A = np.tile(A, (receivers_per_node, transmitters_per_node))
+
+        node_to_receivers = {n: [node_to_int[n] + i*len(node_to_int) for 
+            i in range(receivers_per_node)] for n in node_to_int}
+
+        node_to_transmitters = {n: [node_to_int[n] + i*len(node_to_int) for 
+            i in range(transmitters_per_node)] for n in node_to_int}
+
+    return A, node_to_transmitters, node_to_receivers
+
+def create_channel(num_receivers: int = 1, 
+                   num_transmitters: int = 1,
+                   F_distribution: Optional[Tuple[str, str]] = None,
+                   random_state: Optional[Union[int, np.random.mtrand.RandomState]] = None,
+                   attenuation_matrix: Optional[np.ndarray] = None) -> Tuple[
+                       np.ndarray, float, np.random.mtrand.RandomState]:
```

Comment:
The `typing` claims this returns a 3-tuple, but the docs and code return a 2-tuple.